java-checks/src/main/java/org/sonar/java/checks/spring/SpringAntMatcherOrderCheck.java

```diff
@@ -0,0 +1,185 @@
+      .name("antMatchers")
+      .withAnyParameters());
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return METHOD_MATCHERS;
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree method) {
+    List<StringConstant> patterns = antMatchersPatterns(method);
+    if (!patterns.isEmpty()) {
+      List<StringConstant> previousPatterns = new PreviousMethodDetection().scan(method).previousPatterns;
+      patterns.forEach(pattern -> checkAntPattern(pattern, previousPatterns));
+    }
+  }
+
+  private void checkAntPattern(StringConstant pattern, List<StringConstant> previousPatterns) {
+    for (int i = previousPatterns.size() - 1; i >= 0; i--) {
+      StringConstant previousPattern = previousPatterns.get(i);
+      if (matches(previousPattern.value, pattern.value)) {
+        List<JavaFileScannerContext.Location> secondary = Collections.singletonList(
+          new JavaFileScannerContext.Location("Less restrictive", previousPattern.expression));
+
+        reportIssue(pattern.expression, "Reorder the URL patterns from most to less specific, the pattern \"" +
+          pattern.value + "\" should occurs before \"" + previousPattern.value + "\".", secondary, null);
+        break;
+      }
+    }
+  }
+
+  @VisibleForTesting
+  static boolean matches(String pattern, String text) {
+    if (pattern.equals(text)) {
+      return true;
+    }
+    if (pattern.endsWith("**") && text.startsWith(pattern.substring(0, pattern.length() - 2))) {
+      return true;
+    }
+    boolean antPatternContainsRegExp = pattern.contains("{");
+    boolean textIsAlsoAnAntPattern = MATCHER_SPECIAL_CHAR.matcher(text).find();
+    if (pattern.isEmpty() || antPatternContainsRegExp || textIsAlsoAnAntPattern) {
+      return false;
+    }
+    return text.matches(antMatcherToRegEx(pattern));
+  }
+
+  @VisibleForTesting
+  static String antMatcherToRegEx(String pattern) {
+    // Note, regexp is not supported: {spring:[a-z]+} matches the regexp [a-z]+ as a path variable named "spring"
+
+    // escape regexp special characters
+    return escapeRegExpChars(pattern)
+      // ? matches one character
+      .replace("?", "[^/]")
+      // ** matches zero or more directories in a path ("$$" is a temporary place holder)
+      .replace("**", "$$")
+      // * matches zero or more characters
+      .replace("*", "[^/]*")
+      .replace("$$", ".*");
+  }
+
+  @VisibleForTesting
+  static String escapeRegExpChars(String pattern) {
+    return pattern.replaceAll("([.(){}+|^$\\[\\]\\\\])", "\\\\$1");
+  }
+
+  private static class PreviousMethodDetection {
+
+    List<StringConstant> previousPatterns = new ArrayList<>();
+
+    private PreviousMethodDetection scan(MethodInvocationTree parent) {
+      MethodInvocationTree childMethod = childMethodInvocation(parent);
+      while (childMethod != null) {
```

Comment:
I don't like the fact that for each method invocations, we are going to scan again and again all the previous (child) chained invocations. This could and should be computed only once.